### PR TITLE
ROX-12577 Scanner: load Istio dump

### PR DESCRIPTION
**Pull request by daynewlee**
In this pr, we want to make sure Scanner can load Istio dump, detect changes in Istio dump in run time and in CI.

### CODE REVIEW

File: cmd/updater/diffdumps/cmd.go [hunk 1]
```diff
@@ -99,10 +107,10 @@ func generateK8sDiffs(outputDir string, baseZipR *zip.ReadCloser, headZipR *zip.
 		}
 
 		// Only look at YAML files in the k8s/ folder.
-		if filepath.Dir(name) != vulndump.K8sDirName || filepath.Ext(name) != ".yaml" {
+		if filepath.Dir(name) != dirName || filepath.Ext(name) != ".yaml" {
 			continue
 		}
-		if err := generateK8sDiff(k8sSubDir, baseFiles[name], headF); err != nil {
+		if err := generateK8sDiff(subDir, baseFiles[name], headF); err != nil {
```
Comment (c-du) [inline cmd/updater/diffdumps/cmd.go:113]: It should not be generateK8sDiff here. This function does not seem to be generic to different component.

File: cmd/updater/diffdumps/cmd.go [hunk 2]
```diff
@@ -99,10 +107,10 @@ func generateK8sDiffs(outputDir string, baseZipR *zip.ReadCloser, headZipR *zip.
 		}
 
 		// Only look at YAML files in the k8s/ folder.
```
Comment (c-du) [inline cmd/updater/diffdumps/cmd.go:109]: K8s?

File: cmd/updater/diffdumps/cmd.go (hunk 1 above)
Comment (daynewlee) [reply cmd/updater/diffdumps/cmd.go:113]: Yeah, this is wrong. I will fix this part.

**Review by c-du — APPROVED** [review]

File: pkg/vulnloader/istioloader/yaml.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"io"
+
+	"github.com/ghodss/yaml"
+	"github.com/pkg/errors"
+	"github.com/stackrox/istio-cves/types"
+)
+
+// LoadYAMLFileFromReader loads the Kubernetes CVE feed from the given io.Reader.
```
Comment (RTann) [inline pkg/vulnloader/istioloader/yaml.go:11]: Istio vulnerability feed

File: pkg/vulnloader/istioloader/yaml.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+// It does NOT close the reader; that is the caller's responsibility.
+func LoadYAMLFileFromReader(r io.Reader) (types.Vuln, error) {
+	contents, err := io.ReadAll(r)
+	if err != nil {
+		return types.Vuln{}, errors.Wrap(err, "reading YAML contents")
+	}
+	var schema types.Vuln
```
Comment (RTann) [inline pkg/vulnloader/istioloader/yaml.go:18]: `schema` => `vuln`

File: image/scanner/dump/genesis_manifests.json [hunk 5]
```diff
@@ -308,45 +331,53 @@
...
       "dumpLocationInGS": "gs://stackrox-scanner-ci-vuln-dump/genesis-20220707001950.zip",
       "timestamp": "2022-07-07T00:19:50.023423474Z",
-      "uuid": "b36af074-8f40-4221-a4f6-b96d05d177dd"
+      "uuid": "b36af074-8f40-4221-a4f6-b96d05d177dd",
+      "config": {
+        "ignoreIstioVulns": true
+      }
     },
     {
       "dumpLocationInGS": "gs://stackrox-scanner-ci-vuln-dump/genesis-20220908183429.zip",
```
Comment (RTann) [inline image/scanner/dump/genesis_manifests.json:383]: It should be added here, too

File: e2etests/testcase_test.go [hunk 6]
```diff
@@ -1990,7 +1990,7 @@ var testCases = []testCase{
 				},
 				AddedBy:  "sha256:36e8e9714b9a509fae9e515ff16237928c3d809f5ae228b14d2f7d7605c02623",
 				Location: "jars/jackson-databind-2.9.10.4.jar",
-				FixedBy:  "2.12.6.1",
+				FixedBy:  "2.14.0",
```
Comment (RTann) [inline e2etests/testcase_test.go:1993]: This should really be done in a separate PR, but it's fine for now

File: cmd/updater/diffdumps/cmd.go [hunk 7]
```diff
@@ -98,12 +148,18 @@ func generateK8sDiffs(outputDir string, baseZipR *zip.ReadCloser, headZipR *zip.
 			continue
 		}
 
-		// Only look at YAML files in the k8s/ folder.
-		if filepath.Dir(name) != vulndump.K8sDirName || filepath.Ext(name) != ".yaml" {
+		// Only look at YAML files in the k8s/ or istio/ folder.
```
Comment (RTann) [inline cmd/updater/diffdumps/cmd.go:151]: in case we add another for some reason, let's just say "Only look at YAML files in the given `dirName ` directory."

File: cmd/updater/diffdumps/cmd.go [hunk 8]
```diff
@@ -98,12 +148,18 @@ func generateK8sDiffs(outputDir string, baseZipR *zip.ReadCloser, headZipR *zip.
...
+		if filepath.Dir(name) != dirName || filepath.Ext(name) != ".yaml" {
 			continue
 		}
-		if err := generateK8sDiff(k8sSubDir, baseFiles[name], headF); err != nil {
-			return errors.Wrapf(err, "generating Kubernetes diff for file %q", headF.Name)
+		if outputDir == vulndump.K8sDirName {
```
Comment (RTann) [inline cmd/updater/diffdumps/cmd.go:155]: Instead of doing this check, maybe we pass in a `generateDiffFunc` to this function since they both have the same function signature

File: cmd/updater/diffdumps/cmd.go [hunk 9]
```diff
@@ -69,9 +71,57 @@ func generateK8sDiff(outputDir string, baseF, headF *zip.File) error {
...
+	}
+
+	outF, err := os.Create(filepath.Join(outputDir, filepath.Base(headF.Name)))
+	if err != nil {
+		return errors.Wrap(err, "creating Istio output file")
+	}
+	defer utils.IgnoreError(outF.Close)
+
+	if !reflect.DeepEqual(baseIstioDump, istioDump) {
+		log.Infof("Kubernetes CVE file %q is in the diff", headF.Name)
```
Comment (RTann) [inline cmd/updater/diffdumps/cmd.go:105]: Istio

File: cmd/updater/diffdumps/cmd.go [hunk 10]
```diff
@@ -69,9 +71,57 @@ func generateK8sDiff(outputDir string, baseF, headF *zip.File) error {
...
+		if _, err := io.Copy(outF, headReader); err != nil {
+			return errors.Wrap(err, "copying Kubernetes CVE file")
```
Comment (RTann) [inline cmd/updater/diffdumps/cmd.go:107]: Istio vulnerability

File: e2etests/testcase_test.go (hunk 6 above)
Comment (daynewlee) [reply e2etests/testcase_test.go:1993]: Yeah, I was using another pr but the CI always timed out. Meanwhile I didn't remove this change from this pr.

File: cmd/updater/diffdumps/cmd.go [hunk 11]
```diff
@@ -98,12 +150,12 @@ func generateK8sDiffs(outputDir string, baseZipR *zip.ReadCloser, headZipR *zip.
...
-		// Only look at YAML files in the k8s/ folder.
-		if filepath.Dir(name) != vulndump.K8sDirName || filepath.Ext(name) != ".yaml" {
+		// Only look at YAML files in the <dirName> folder.
+		if filepath.Dir(name) != dirName || filepath.Ext(name) != ".yaml" {
 			continue
 		}
-		if err := generateK8sDiff(k8sSubDir, baseFiles[name], headF); err != nil {
-			return errors.Wrapf(err, "generating Kubernetes diff for file %q", headF.Name)
+		if err := generateDiffs(subDir, baseFiles[name], headF); err != nil {
+			return errors.Wrapf(err, "generating Istio diff for file %q", headF.Name)
```
Comment (RTann) [inline cmd/updater/diffdumps/cmd.go:158]: remove `Istio`, as it's not clear if this is k8s or Istio at this time

**Review by RTann — COMMENTED** [review]
very minor nits. Did you also confirm the vulns are stored in the `scanner` image as expected?

**Review by RTann — APPROVED** [review]
Don't forget to change the PR title :)
